manimlib/mobject/number_line.py

```diff
@@ -10,77 +10,83 @@
 from manimlib.mobject.types.vectorized_mobject import VGroup
 from manimlib.utils.bezier import interpolate
 from manimlib.utils.bezier import outer_interpolate
-from manimlib.utils.config_ops import digest_config
-from manimlib.utils.config_ops import merge_dicts_recursively
+from manimlib.utils.dict_ops import merge_dicts_recursively
 from manimlib.utils.simple_functions import fdiv
 
 from typing import TYPE_CHECKING
 
 if TYPE_CHECKING:
-    from typing import Iterable, Sequence
+    from typing import Iterable
+    from manimlib.typing import ManimColor, Vect3, Vect3Array, VectN, RangeSpecifier
 
 
 class NumberLine(Line):
-    CONFIG = {
-        "color": GREY_B,
-        "stroke_width": 2,
-        # List of 2 or 3 elements, x_min, x_max, step_size
-        "x_range": [-8, 8, 1],
+    def __init__(
+        self,
+        x_range: RangeSpecifier = (-8, 8, 1),
+        color: ManimColor = GREY_B,
+        stroke_width: float = 2.0,
         # How big is one one unit of this number line in terms of absolute spacial distance
-        "unit_size": 1,
-        "width": None,
-        "include_ticks": True,
-        "tick_size": 0.1,
-        "longer_tick_multiple": 1.5,
-        "tick_offset": 0,
+        unit_size: float = 1.0,
+        width: float | None = None,
+        include_ticks: bool = True,
+        tick_size: float = 0.1,
+        longer_tick_multiple: float = 1.5,
+        tick_offset: float = 0.0,
         # Change name
-        "numbers_with_elongated_ticks": [],
-        "include_numbers": False,
-        "line_to_number_direction": DOWN,
-        "line_to_number_buff": MED_SMALL_BUFF,
-        "include_tip": False,
-        "tip_config": {
-            "width": 0.25,
-            "length": 0.25,
-        },
-        "decimal_number_config": {
-            "num_decimal_places": 0,
-            "font_size": 36,
-        },
-        "numbers_to_exclude": None
-    }
-
-    def __init__(self, x_range: Sequence[float] | None = None, **kwargs):
-        digest_config(self, kwargs)
-        if x_range is None:
-            x_range = self.x_range
-        if len(x_range) == 2:
-            x_range = [*x_range, 1]
-
-        x_min, x_max, x_step = x_range
-        # A lot of old scenes pass in x_min or x_max explicitly,
-        # so this is just here to keep those workin
-        self.x_min: float = kwargs.get("x_min", x_min)
-        self.x_max: float = kwargs.get("x_max", x_max)
-        self.x_step: float = kwargs.get("x_step", x_step)
-
-        super().__init__(self.x_min * RIGHT, self.x_max * RIGHT, **kwargs)
-        if self.width:
-            self.set_width(self.width)
-            self.unit_size = self.get_unit_size()
+        numbers_with_elongated_ticks: list[float] = [],
+        include_numbers: bool = False,
+        line_to_number_direction: Vect3 = DOWN,
+        line_to_number_buff: float = MED_SMALL_BUFF,
+        include_tip: bool = False,
+        tip_config: dict = dict(
```

Comment:
I noticed there are many places using dicts (mutable) as default arguments. It could be better to convert them into None and set the default values accordingly.

Reply:
I agree. This initial pass was mainly geared at converting CONFIG state to __init__ parameters, fairly directly, but I'll do another sweep for things like that.